sappho/collide.py

```diff
@@ -1,11 +1,402 @@
+                # the width of 11.
+                this_partitions_width += last_partition_extra_width
+
+            # The calculation below tells us if we're on the last
+            # row or not, in order to compensate/stretch.
+            if partition_index >= area_in_partitions - partitions_wide:
+                this_partitions_height += last_partition_extra_height
+
+            # Finally, we've gotten all the info required to accurately
+            # create this spacial partition and add it to the list of
+            # spatial partitions to return!
+            spatial_partition = SpatialPartition(
+                x=partition_x,
+                y=partition_y,
+                width=this_partitions_width,
+                height=this_partitions_height,
+            )
+            list_of_spatial_partitions.append(spatial_partition)
+
+        return SpatialPartitionGrid(
+            list_of_spatial_partitions,
+        )
+
+    def pixel_coordinates_to_partition(self, x, y):
+        """
+
+        Examples:
+            To continue the example...
+
+            >>> grid = SpatialPartitionGrid.from_dimensions(
+            ...     pixels_wide=31,
+            ...     pixels_tall=41,
+            ...     partitions_wide=3,
+            ...     partitions_tall=4,
+            ... )
+            >>> lol = grid.pixel_coordinates_to_partition(x=25, y=34)
+            >>> lol is grid._partitions[-1]
+            True
+
+        """
+
+        new_x = x // self.normal_partition_width
+        new_y = y // self.normal_partition_height
+        index = (new_x + (new_y * self.width_in_partitions))
+        return self._partitions[index]
+
+    # FIXME
+    def intersecting_partitions(self, pygame_rect):
+        """Return the partitions which a rectangle occupies.
+
+        Warning:
+            Since this just checks a rectangles's four
+            corners, it will not work properly if the
+            rectangle is greater than two partitions in size.
+
+        Arguments:
+            pygame_rect (pygame.Rect): ...
+
+        Returns:
+            set(SpatialPartition): ...
+
+        Examples:
+            >>> grid = SpatialPartitionGrid.from_dimensions(
+            ...     pixels_wide=31,
+            ...     pixels_tall=41,
+            ...     partitions_wide=3,
+            ...     partitions_tall=4,
+            ... )
+            >>> rect = pygame.Rect(15, 25, 10, 10)
+            >>> rect.topleft
+            (15, 25)
+            >>> parts = grid.intersecting_partitions(rect)
+            >>> len(parts)
+            4
+
+        """
+
+        if ((pygame_rect.width > self.normal_partition_width * 2)
+           or (pygame_rect.height > self.normal_partition_height * 2)):
+            # update this part to get the partitions in between, in the future
+            message = 'Rect must not exceed 2x height nor width. Support soon!'
+            raise NotImplementedError(message)
+        else:
+            return set([
```

Comment:
um... `SpatialPartition` isn't a serializable object...

could just use a long string of conditions keep same complexity